content/blog/updates/2024-01-09-incident/index.md

```diff
@@ -0,0 +1,84 @@
+
+{{< figure src="total-databases.png" alt="A graph of total databases per CouchDB node. We see 3 nodes sitting the same value, and at 13:52 a new nodes joins and the number of databases on it starts to increase slowly" >}}
+
+Our production database cluster is called "chesterfield" (after the [sofa][3]), and each of the nodes is numbered sequentially. `004` is the new node in this graph, and you can see when it joins, when it starts getting data, when we disconnect it from the cluster, and when we bring it back down.
+
+The mistake we made in our automated process for adding nodes was to add the new node to our load balancer before it had fully synchronised. You can see that in this graph showing the number of healthy nodes our database load balancer is aware of.
+
+{{< figure src="healthy-nodes.png" alt="A graph showing a line that hovers at the value 3 until 13:52, where it jumps to 4, and then at 13:58 goes back down to 3 and stays there" >}}
+
+CouchDB is what's called a "master-master" database. Many distributed database systems will have one node that's more important
+than the others (the "master" node) and this is the node responsible for handling new data, while the other nodes can only handle the fetching of data. This is not the case in CouchDB, all nodes are equal and can handle writes. The way this works under the hood is that conflicting writes, e.g. 2 separate writes to the same data on 2 different nodes, is resolved through a conflict-resolution process.
+
+For historical reasons, most requests that Budibase makes to its internal database first check to make sure that the database they're writing to exists. If it does not, it gets created. For the 6 minutes that `chesterfield-004` was in the load balancer, most requests to check that a database existed would have reported that no, the database does not exist. A new, empty database would be created and the code would continue. This empty database was then replicated to other nodes, effectively deleting all of their contents. We were able to reproduce this behaviour locally, and you can check our methodology [here][4].
+
+{{< figure src="chesterfield-004-writes.png" alt="A graph showing the number of writes per second to chesterfield-004. It peaks at 1 write per second at 13:54, then trails off until the line disappears at 13:58">}}
+
+This manifested for most users as the inability to log in, as their user record had been deleted.
+
+{{< figure src="login-403s.png" alt="A graph showing HTTP requests to Budibase Cloud that had a 403 response, indicating login failure" >}}
+
+Unfortunately we were not aware that there was a problem until 14:13 UTC, when we started receiving reports from customers that they were not able to log in. Why, then, had we removed the node from the cluster and turned it down?
```

Comment:
Possibly worth mentioning here that we do have automation to check for this, but it didn't trigger because it only affected a certain amount of customers

Reply:
Ooh, what do we have for this?

Reply:
There's a playwright suite that runs and checks login every few minutes. It runs here:
https://github.com/Budibase/budibase-devops/actions/runs/7532366712/job/20502803167

Reply:
I've added a section mentioning this 🙏